Doc on clusterctl. How to use and how it works. (#1131)
kubernetes-sigs/cluster-api: docs/how-to-use-clusterctl.md

## Using `clusterctl` to create a cluster from scratch

> This document provides an overview of how `clusterctl` works and explains how one can use `clusterctl` to create a kubernetes cluster from scratch.

Reviewer: I would capitalize Kubernetes everywhere

### What is `clusterctl`?

> `clusterctl` is a CLI tool to create a kubernetes cluster. It uses Cluster API provider implementations to provision resourses needed by the kubernetes cluster.

Reviewer 2: There should probably be a statement here to let users know that `clusterctl` is provided by provider implementations and not a single common tool.

Author: Good point. I will add a statement to make it clear.

Reviewer: Typo: resourses should be resources.

### Creating a cluster

> `clusterctl` needs 4 YAML files to start with: `provider-components.yaml`, `cluster.yaml`, `machines.yaml` , `addons.yaml`.  

Reviewer 2: Technically there is a 5th optional file that can be used with the `--bootstrap-only-components` to provide resources to the bootstrap cluster without also providing them to the target cluster post-pivot.

Author: I didn't use this option much. Can you provide an example use case for this? I will find a good place to add it to the doc.

Reviewer 2: I don't have an example offhand, but it's use would be similar to the flags used for the mentioned yaml files and could contain `Secrets` or other resources that would only be needed in the bootstrap provider.

* `addons.yaml` contains the addons for the provider.

Many providers implemenatations come with helpful scripts to generate these YAMLS. Provider implementation can be found [here](https://github.com/kubernetes-sigs/cluster-api#provider-implementations).  

After generating the YAML files run the following command: 
```
clusterctl create cluster --provider <PROVIDER> --bootstrap-type <BOOTSTRAP CLUSTER TYPE> -c cluster.yaml -m machines.yaml -p provider-components.yaml --addon-components addons.yaml
```
Example usage:
```
# VMWare VSphere
clusterctl create cluster --provider vsphere --bootstrap-type kind -c cluster.yaml -m machines.yaml -p provider-components.yaml --addon-components addons.yaml

# Amazon AWS
clusterctl create cluster --provider aws --bootstrap-type kind -c cluster.yaml -m machines.yaml -p provider-components.yaml --addon-components addons.yaml
```

**What happens when we run the command?**  
After running the command first it creates a local cluster. If `kind` was passed as the `--bootstrap-type` it creates a local [kind](https://kind.sigs.k8s.io/) cluster. This cluster is generally referred to as the *bootstrap cluster*.  

> On this kind kubernetes cluster the `provider-components.yaml` file is applied. This step loads the CRDs into the cluster. It also creates 2 [StatefulSet](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/) pods that run the cluster api controller and the provider specific controller. These pods register the custom controllers that manage the newly defined resources (`Cluster`, `Machine`, `MachineSet`, etc) with the [kubernetes controller manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/).  

Reviewer: Everything is accurate here except the part about the kubernetes controller manager. That is not involved here. The pods themselves contain the controllers that manage the Cluster/MachineDeployment/MachineSet/Machine custom resources.

  
Next, `clusterctl` applies the `cluster.yaml` and `machines.yaml` to the local kind kubernetes cluster. This step creates a kubernetes cluster with only a control-plane(as defined in `machines.yaml`) on the specified provider. This newly created cluster is generally referred to as the *management cluster* or *pivot cluster* or *initial target cluster*. The management cluster is resonsible for creating and maintaining the work-load cluster.  
  
Lastly, `clusterctl` moves all the CRDs and the custom controllers from the bootstrap cluster to the management cluster and deletes the locally created bootstrap cluster.  

### Creating a workload cluster using the management cluster

> The *workload cluster* also sometimes referred to as the *target cluster* is the kubernetes custer on to which the final application is deployed. The target cluster is responsible for handling the workload of the application, not the management cluster.

Reviewer 2: ```suggestion
The *workload cluster* also sometimes referred to as the *target cluster* is the Kubernetes cluster on to which the final application is deployed. The target cluster is responsible for handling the workload of the application, not the management cluster.
```

> As the management cluster is up we can create a workload cluster by simply applying the appropriate `cluster.yaml`, `machines.yaml` and `machineset.yaml` on the management cluster. This will create the VMs(Nodes) as defined in these YAMLs and uses `kubeadm` (generally pre-installed on these VMs) to perform `init` and `join` to create a kubernetes cluster on them.  

Reviewer 2: This statement is true for the majority of providers, but `kubeadm` isn't a required implementation detail.

Author: I thought `kubeadm` was the only way. Guess not.
What do VMs that don't use `kubeadm` use to create the cluster? I would be better if we add a small line about the non-kubeadm way as well.

Reviewer 2: It could really be anything. The provider could deploy a script for a Kubernetes distribution, such as OpenShift or Rancher, or it could even potentially implement Kubernetes The Hard Way. Today all of the providers I am aware of use `kubeadm`, but it isn't a hard requirement.

# Using `clusterctl` to create a cluster from scratch

This document provides an overview of how `clusterctl` works and explains how one can use `clusterctl`
to create a Kubernetes cluster from scratch.

## What is `clusterctl`?

`clusterctl` is a CLI tool to create a Kubernetes cluster. `clusterctl` is provided by the [provider implementations](https://github.com/Kubernetes-sigs/cluster-api#provider-implementations).
It uses Cluster API provider implementations to provision resources needed by the Kubernetes cluster.

## Creating a cluster

`clusterctl` needs 4 YAML files to start with: `provider-components.yaml`, `cluster.yaml`, `machines.yaml` ,
`addons.yaml`.

> * `provider-components.yaml` constain the *Custom Resource Definitions ([CRDs](https://Kubernetes.io/docs/concepts/extend-Kubernetes/api-extension/custom-resources/))* 

Reviewer: ```suggestion
* `provider-components.yaml` contains the *Custom Resource Definitions ([CRDs](https://Kubernetes.io/docs/concepts/extend-Kubernetes/api-extension/custom-resources/))* 
```

of all the resources that are managed by Cluster API. Some examples of these resources
are: `Cluster`, `Machine`, `MachineSet`, etc. For more details about Cluster API resources
click [here](https://cluster-api.sigs.k8s.io/common_code/architecture.html#cluster-api-resources).
* `cluster.yaml` defines an object of the resource type `Cluster`.
* `machines.yaml` defines an object of the resource type `Machine`. Generally creates the machine
that becomes the control-plane.
* `addons.yaml` contains the addons for the provider.

> Many providers implemenatations come with helpful scripts to generate these YAMLS. Provider implementation

Reviewer: ```suggestion
Many providers implementations come with helpful scripts to generate these YAML files. Provider implementations
```

can be found [here](https://github.com/Kubernetes-sigs/cluster-api#provider-implementations).  

`clusterctl` also comes with additional features. Example, `clusterctl` can also take in an optional
`bootstrap-only-components.yaml` to provide resources to the bootstrap cluster without also providing them
to the target cluster post-pivot.

> For more details about all the supported options run:

Reviewer: Add a blank line above this one?

```
clusterctl create cluster --help
```

After generating the YAML files run the following command:

```
clusterctl create cluster --provider <PROVIDER> --bootstrap-type <BOOTSTRAP CLUSTER TYPE> -c cluster.yaml -m machines.yaml -p provider-components.yaml --addon-components addons.yaml
```

Example usage:

```

> # VMWare VSphere

Reviewer: ```suggestion
# VMware vSphere
```

On this kind Kubernetes cluster the `provider-components.yaml` file is applied. This step loads the CRDs into
the cluster. It also creates 2 [StatefulSet](https://Kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
pods that run the cluster api controller and the provider specific controller. These pods register the custom
controllers that manage the newly defined resources (`Cluster`, `Machine`, `MachineSet`, etc).  

Next, `clusterctl` applies the `cluster.yaml` and `machines.yaml` to the local kind Kubernetes cluster. This
step creates a Kubernetes cluster with only a control-plane(as defined in `machines.yaml`) on the specified
provider. This newly created cluster is generally referred to as the *management cluster* or *pivot cluster*
or *initial target cluster*. The management cluster is responsible for creating and maintaining the work-load cluster.  
  
Lastly, `clusterctl` moves all the CRDs and the custom controllers from the bootstrap cluster to the
management cluster and deletes the locally created bootstrap cluster. This step is referred to as the *pivot*.

### Creating a workload cluster using the management cluster

> The *workload cluster* also sometimes referred to as the *target cluster* is the Kubernetes custer on to which

Reviewer 2: ```suggestion
The *workload cluster* also sometimes referred to as the *target cluster* is the Kubernetes cluster on to which
```

the final application is deployed. The target cluster is responsible for handling the workload of the application,
not the management cluster.

As the management cluster is up we can create a workload cluster by simply applying the appropriate

> `cluster.yaml`, `machines.yaml` and `machineset.yaml` on the management cluster. This will create the VMs(Nodes)

Reviewer: You also need to apply addons for a CNI plugin; otherwise, your nodes will be `NotReady`.

It uses Cluster API provider implementations to provision resources needed by the Kubernetes cluster.

## Creating a cluster

`clusterctl` needs 4 YAML files to start with: `provider-components.yaml`, `cluster.yaml`, `machines.yaml` ,
`addons.yaml`.

* `provider-components.yaml` contains the *Custom Resource Definitions ([CRDs](https://Kubernetes.io/docs/concepts/extend-Kubernetes/api-extension/custom-resources/))* 
of all the resources that are managed by Cluster API. Some examples of these resources
are: `Cluster`, `Machine`, `MachineSet`, etc. For more details about Cluster API resources
click [here](https://cluster-api.sigs.k8s.io/common_code/architecture.html#cluster-api-resources).
* `cluster.yaml` defines an object of the resource type `Cluster`.
* `machines.yaml` defines an object of the resource type `Machine`. Generally creates the machine
that becomes the control-plane.
* `addons.yaml` contains the addons for the provider.

Many providers implementations come with helpful scripts to generate these YAMLS. Provider implementation
can be found [here](https://github.com/Kubernetes-sigs/cluster-api#provider-implementations).  

> `clusterctl` also comes with additional features. Example, `clusterctl` can also take in an optional

Reviewer: ```suggestion
`clusterctl` also comes with additional features. For example, `clusterctl` can also take in an optional
```

`bootstrap-only-components.yaml` to provide resources to the bootstrap cluster without also providing them
to the target cluster post-pivot.

For more details about all the supported options run:

```
clusterctl create cluster --help
```

After generating the YAML files run the following command:

```
clusterctl create cluster --provider <PROVIDER> --bootstrap-type <BOOTSTRAP CLUSTER TYPE> -c cluster.yaml -m machines.yaml -p provider-components.yaml --addon-components addons.yaml
```

Example usage:

```

> # VMWare vSphere

Reviewer: ```suggestion
# VMware vSphere
```
